### ISSUE TITLE
NVDA accidentally deleted a user's private files while creating a portable copy

### ISSUE THREAD

**Issue by cary-rowen** [nvaccess/nvda#16514]
## background
In the NVDA Chinese community, a user reported an issue: while creating a portable copy of NVDA, they accidentally selected the root directory of the D drive as the target location, due to overlooking the fact that NVDA should be placed in a separate directory.

Below is a screenshot of a post made by a user on the BBS.

![image](https://github.com/nvaccess/nvda/assets/21048245/a5845763-33fa-4133-8ba9-1f55bf2a633b)

I have translated this user's Chinese post into English:

> "Big trouble here! I was using the latest version of NVDA, 2024.1, to create a portable version. I wasn’t paying attention and just chose the D drive root directory as the location without creating a new folder. I also checked the option to copy the configuration files. There were no warnings at all, and then the creation process started! It took a really long time, maybe about a minute or more. After it was done, I opened 'This PC' and saw nothing but panes. I restarted Windows Explorer without noticing anything wrong at first. However, when I tried to open applications installed on the D drive, they all failed, reporting missing DLLs. I checked, and it turns out, every single DLL file was gone. It wiped out everything in every folder, sparing nothing—not even the corners. It deleted all my personal data, including some DLLs I had stored. So, this portable creation feature of NVDA is deadly! It wrecked my D drive without any warning. Now, I’m asking if there's any way to recover the data deleted by NVDA, though I know it’s a long shot. Also, I want to warn everyone not to choose any root directory as the location for creating the NVDA portable version. This has really done me in! I definitely need to reinstall my operating system now, and then try to recover my personal data. So annoying!"

## some investigation

**I read the source code of NVDA and found the cause of the problem**

[nvaccess/nvda/blob/master/source/installer.py#L230-L237](https://github.com/nvaccess/nvda/blob/1791eb96a41778dfd4a93715bb95db7481b1a3b9/source/installer.py#L230-L237)

In simple terms, NVDA recursively removes files with specific extensions within the target directory and all its subdirectories.

Those removed include: ".pyc", ".pyo", ".pyd", ".dll", ".manifest".

From the context, it's evident that because the user set the D drive root directory as the target for creating the NVDA portable copy, the subsequent execution of the code led to the deletion of all files with the specified extensions that the user had. This resulted in the malfunction of software installed on the D drive and even the deletion of the user's valuable files.

I am hesitant to label this as the same category of issues as #3113 and #16443.
No matter the circumstances, the deletion of users' private files should not occur. It's important to note that setting the D drive as the target directory for the portable copy is not inherently a user error; in fact, it is permissible.

### Steps to reproduce:
1. Create an empty directory named "nvda" in the root of the D drive.
2. Place `test.dll`, `test.txt`, and `test.manifest` in this directory.
3. Inside `d:\nvda`, create a new directory named "test" and place `test.dll`, `test.txt`, and `test.manifest` in this newly created "test" directory.
4. Create a portable copy of NVDA with the target directory set to `d:\nvda`.

### Actual behavior:
After step (4) is completed, only `test.txt` remains; other files have been deleted.

### Expected behavior:
These directories and files are unrelated to NVDA and are personal files of the user; they should not be deleted.

### NVDA logs, crash dumps, and other attachments:
n/a

#### NVDA installed/portable/running from source:
Installed

#### NVDA version:
Potentially all versions

#### Windows version:
n/a
#### Name and version of other software in use when reproducing the issue:
n/a
#### Other information about your system:
n/a
#### Does the issue still occur after restarting your computer?
yes
#### Have you tried any other versions of NVDA? If so, please report their behaviors.
n/a
#### If NVDA add-ons are disabled, is your problem still occurring?
yes
#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?
n/a

**Comment by cary-rowen** [comment]
cc @gerald-hartig @seanbudd 
I'd recommend giving this a higher priority or classifying it as a security flaw.

**Comment by gerald-hartig** [comment]
I understand why we might want to install into root directories, for portable USB drives with only NVDA on it. However the first comment in the screenshot saying "Indeed, I have encountered this before" frightens me a little.

I see four things we could do here:

1. Block installation on root directories or warn the user.
1. Check how many files are going to be deleted and if the number is greater than expected then warn the user.
1. Check the names of the files and if they match a blacklist glob (or don't fit in the whitelist glob) then warn the user.
1. Check if the drive is not a removable drive (`psutil.disk_partitions() check partition.opts for 'removable'`) then warn the user.

**Comment by cary-rowen** [comment]
Hi @gerald-hartig 

The screenshot a just part of the post, which contains dozens of responses so far, with many more people saying they've been in similar situations.


You wrote:
> 1. Block installation on root directories or warn the user.

I don't think the cause of the problem is that drive root or not, NVDA recursively deletes all files with a specific extension in the target directory and its subdirectories, even if the directory is completely unrelated to NVDA.

> 3. Check the names of the files and if they match a blacklist glob (or don't fit in the whitelist glob) then warn the user.

If you just blacklist the file name, there are also security risks:

Let's say I have NVDA directory on my D drive, which stores 23.1, 23.2, 23.3, which are portable copies of different versions of NVDA.
When I tried to put NVDA2024.1 into d:\NVDA, these files that were included in the deletion list were also included in its subdirectories (portable copies of NVDA with different versions). As you can imagine, other copies of NVDA will be Total destruction.
Unless we add more conditional checks or stricter filename matching. this is not elegant.

> 4. Check if the drive is not a removable drive ( psutil.disk_partitions() check partition.opts for 'removable' ) then warn the user.

No, even a removable drive does not mean that the NVDA user only wants to store a portable copy of NVDA on this removable drive, more often the user stores more other portable programs, the problem still occurs and is more Oops.

You mentioned "warn the user". However, IMHO: all warnings may be ignored by users. For us, this is only legal. This is just like many users who do not read the software license text carefully before installing the software.

A typical example:
I get a lot of questions from users, why are all add-ons disabled after I upgraded to NVDA2024.1?
In fact, users had already checked the box before upgrading that add-ons would be disabled, but they seemed to have completely forgotten about it.
Another example is overriding compatibility warnings and installing add-ons that are not compatible at all.

**Comment by thgcode** [comment]
An idea, some programs require to be installed on an empty folder, maybe NVDA could check if the folder contains any file and ask the user to create an empty folder to install itself into, or it could create an empty folder itself.

**Comment by gerald-hartig** [comment]
What if we only install on either empty directories or something that is clearly the root of an NVDA installation. We would need a heuristic that looks for specific folders and files present in the root of the installation. 

However, if someone installs by mistake on D:\ and then later upgrades that install, then this wouldn't work. It would happily go ahead and continue down the tree, unless we also checked that the right number of files are being deleted.

**Comment by cary-rowen** [comment]
@thgcode 

> An idea, some programs require to be installed on an empty folder, maybe NVDA could check if the folder contains any file and ask the user to create an empty folder to install itself into, or it could create an empty folder itself.

This is exactly what #3113 and #16643 are discussing.

**Comment by gerald-hartig** [comment]
FYI https://github.com/nvaccess/nvda/issues/16443 not 16643

**Comment by cary-rowen** [comment]
Thanks @gerald-hartig, I fixed it.

**Comment by seanbudd** [comment]
Closing in favour of #16443

We've raised priority to this for a p1, meaning we aim to address it in the current release cycle (ASAP for 2024.3 release).

**Comment by lukaszgo1** [comment]
@seanbudd  While #16443 will certainly improve things, there is more we can do here:
- NVDA should not cleanup folders unless they contained NVDA installs - we have code which detects if the given directory contains NVDA files it just has to be used
- NVDA should not delete `.dll` files recursively - I have implemented this in #10224 and believe it worked back then. Note that there is no technical reason to remove these files in directories other than the root folder.


I can create separate issues for these two items if you prefer.

**Comment by cary-rowen** [comment]
Hi @seanbudd 


I hope to keep this issue open for a while:
1. The affected user is still looking for a solution to recover files. It has some deleted files that are difficult to find. This is not a wish that NV Access should be held responsible, just highlighting the seriousness of the problem.
2. The original description and comments of Issue #16443 did not mention the risk of deleting users' private files. This suggests that Issue #16443 wasn't intended to address the specific use case presented in this issue.
3. In Issue #16443, Sean wrote, "if this is an existing NVDA folder it will upgrade safely." However, 
This problem would still exist if we allowed users to choose not to create NVDA directories.


## In summary, I believe keeping this issue open offers several benefits:
1. Discussion of safer solutions: This can help prevent similar data loss in the future.
2. Helps remind community users that there is a known risk in NVDA versions that have not yet provided a fix.

Thanks to NV Access for the quick response.

Best,
Cary

**Comment by seanbudd** [comment]
@cary-rowen  #16443 has been updated to consider this issue. We regularly merge issues that describe the same base problem or lead to the same base solution.
If the solution to #16433 fails to capture the problem outlined here - even though #15433 has been updated to ensure it encapsulates both these issues, I would consider re-opening this issue.

To be clear closing this issue doesn't prevent discussions on fixing this issue - it just suggests to take that conversation to #16443.
NVDA GitHub issues do not serve as community reminders, it is for logging information and communicating with developers. Communicating the current risk is best done in user centric groups, not the NVDA development hub.

**Comment by seanbudd** [comment]
>NVDA should not cleanup folders unless they contained NVDA installs - we have code which detects if the given directory contains NVDA files it just has to be used.

@lukaszgo1 - is this not sufficiently covered by the warning described in  https://github.com/nvaccess/nvda/issues/16443#issuecomment-2106811009 ?
